flows/engine/session.go

```diff
@@ -51,117 +60,302 @@ func (s *session) GetRun(uuid flows.RunUUID) (flows.FlowRun, error) {
+	for _, run := range s.runs {
+		if run.Status() == flows.RunStatusWaiting {
+			return run
+		}
 	}
 	return nil
 }
 
-func (s *session) addRun(run flows.FlowRun) {
-	s.runs = append(s.runs, run)
-	s.runsByUUID[run.UUID()] = run
-}
-
 func (s *session) LogEvent(step flows.Step, action flows.Action, event flows.Event) {
 	s.log = append(s.log, NewLogEntry(step, action, event))
 }
 func (s *session) Log() []flows.LogEntry { return s.log }
 func (s *session) ClearLog()             { s.log = nil }
 
-// StartFlow starts the flow for the passed in contact, returning the created FlowRun
-func (s *session) StartFlow(flowUUID flows.FlowUUID, parent flows.FlowRun, callerEvents []flows.Event) error {
-	flow, err := s.assets.GetFlow(flowUUID)
+//------------------------------------------------------------------------------------------
+// Flow execution
+//------------------------------------------------------------------------------------------
+
+// StartFlow beings execution of the given flow in this session
+func (s *session) StartFlow(flowUUID flows.FlowUUID, callerEvents []flows.Event) error {
+	flow, _ := s.Assets().GetFlow(flowUUID)
+
+	s.Trigger(flow, nil)
+
+	// off to the races...
+	return s.continueUntilWait(nil, noDestination, nil, callerEvents)
+}
+
+// Resume resumes a waiting session
+func (s *session) Resume(callerEvents []flows.Event) error {
+	if s.status != flows.SessionStatusWaiting {
+		return utils.NewValidationErrors("only waiting sessions can be resumed")
+	}
+
+	// figure out where (i.e. run and step) we began waiting on
+	waitingRun := s.waitingRun()
+	step, _, err := waitingRun.PathLocation()
 	if err != nil {
 		return err
 	}
 
-	// create our new run
-	run := s.CreateRun(flow, parent)
+	// apply our caller events to this step
+	for _, event := range callerEvents {
+		waitingRun.ApplyEvent(step, nil, event)
+	}
+
+	// see if the wait is now satisfied and will let us resume
+	if s.wait.CanResume(waitingRun, step) {
+		s.status = flows.SessionStatusActive
+		waitingRun.SetStatus(flows.RunStatusActive)
+
+		destination, err := s.resumeRun(waitingRun)
+		if err != nil {
+			return err
+		}
 
-	// no first node, nothing to do (valid but weird)
-	if len(flow.Nodes()) == 0 {
-		run.Exit(flows.StatusCompleted)
-		return nil
+		// off to the races again...
+		if err = s.continueUntilWait(waitingRun, destination, step, []flows.Event{}); err != nil {
+			return err
+		}
 	}
 
-	// off to the races
-	return continueRunUntilWait(run, flow.Nodes()[0].UUID(), nil, callerEvents)
+	return nil
 }
 
-// ResumeFlow resumes our session from the last step
-func (s *session) Resume(callerEvents []flows.Event) error {
-	// find the active run
-	run := s.ActiveRun()
-	if run == nil {
-		return utils.NewValidationErrors("session: no active run to resume")
+// resumes a run that may have been waiting or a parent paused for a child subflow
+func (s *session) resumeRun(run flows.FlowRun) (flows.NodeUUID, error) {
```

Comment:
This could probably use a better name, pretty confusing having it be `resumeRun` and not actually continuing the run. Maybe something like `findNextDestination` or something?